### ISSUE TITLE
Be precise about "arguments", "parameters", and "variables"

### ISSUE THREAD

**Issue by HertzDevil** [crystal-lang/crystal#10374]
The reference manual and the compiler source use the terms "arguments", "parameters", and "variables" interchangeably, sometimes in a way inconsistent with most other programming languages. There is a clear distinction between the three terms: (quotes taken from [Wikipedia](https://en.wikipedia.org/wiki/Parameter_(computer_programming)))

> The _argument_ in computer science is the actual input expression passed/supplied to a function, procedure, or routine in the invocation/call statement, whereas the _parameter_ is the variable inside the implementation of the subroutine. For example, if one defines the `add` subroutine as `def add(x, y): return x + y`, then `x, y` are parameters, while if this is called as `add(2, 3)`, then `2, 3` are the arguments.

> A parameter is an (unbound) variable, while the argument can be a literal or variable or more complex expression involving literals and variables.

We should follow the same terminology in Crystal to minimize chances of confusion. (The Eiffel convention of "formal arguments" and "actual arguments" seems rather verbose so I'll use the usual meanings.) An example in the reference manual is [Default values and named arguments](https://crystal-lang.org/reference/syntax_and_semantics/default_and_named_arguments.html):

> ### Default ~~values~~ <ins>arguments</ins>
>
> A method can specify default ~~values~~ <ins>arguments</ins> for the last ~~arguments~~ <ins>parameters</ins>: ...
>
> ### Named arguments
>
> All arguments can also be specified, in addition to their position, by their name. For example: ...
>
> When there are many arguments, the order of the names in the invocation doesn't matter, as long as all required ~~arguments~~ <ins>parameters</ins> are covered: ...
>
> When a method specifies a splat <ins>parameter</ins> (explained in the next section), named arguments can't be used <ins>for positional parameters</ins>. The reason is that understanding how arguments are matched becomes very difficult; positional arguments are easier to reason about in this case.

Another place of misuse is [break](https://crystal-lang.org/reference/syntax_and_semantics/break.html):

> `break` can also take ~~a parameter~~ <ins>an argument</ins> which will then be the value that gets returned:

Whereas in the source code, there is `#type_vars`, which represents all three concepts:

* `Crystal::GenericType#type_vars` corresponds to the generic type _parameters_ (`T, N` for `StaticArray` and `T, R` for `Proc`);
* `Crystal::GenericInstanceType#type_vars` is the one that really means _variables_ (`T => Int32, N => 4` for `StaticArray(Int32, 4)` and `T => Tuple(Int32, Bool), R => String` for `Proc(Int32, Bool, String)`);
* `Crystal::Macros::TypeNode#type_vars` returns the _parameters_ on an uninstantiated generic, but _arguments_ on an instantiated generic (after splat collection):
  ```crystal
  {% StaticArray.type_vars %}               # => [T, N]
  {% StaticArray(Int32, 4).type_vars %}     # => [Int32, 4]
  {% Proc.type_vars %}                      # => [T, R]
  {% Proc(Int32, Bool, String).type_vars %} # => [Tuple(Int32, Bool), String]
  ```
* `Crystal::Macros::Generic#type_vars` also corresponds to type _arguments_: (note that uninstantiated generics passed in the same manner result in a `Path` instead)
  ```crystal
  macro foo(t)
    {% t.type_vars %}
  end

  foo StaticArray(Int32, 4)     # => [Int32, 4]
  foo Proc(Int32, Bool, String) # => [Int32, Bool, String]
  ```

The docs changes are relatively easy, followed by compiler error messages and internal names in the compiler. Names that appear in the macro land are the hardest to replace, and require proper deprecation, but IMO we should do this in the future even if they don't make it for 1.0. For the particular case of `#type_vars`, perhaps we could:

* Rename `Crystal::GenericType#type_vars` to `#type_params`;
* Rename `Crystal::Macros::Generic#type_vars` to `#type_args`;
* Make `Crystal::Macros::TypeNode#type_params` always return the generic type parameters, even on instantiated generics;
* Make `Crystal::Macros::TypeNode#type_vars` not return anything on uninstantiated generics, or expose `#type_args` here to return a `HashLiteral`. (This would also allow us to remove the hacks in `Tuple.new` and `Slice#[]`.)

And `Crystal::Macros::Arg` should become `Param`. This affects `#class_name` directly and I don't know if a deprecation route is possible there.

**Comment by straight-shoota** [comment]
Sounds good.

However, I'm not sure about the particular first example:

> A method can specify default arguments for the last parameters.

This feels odd to me. It should be "default values for the last parameters". They're not arguments passed to the method. The parameters have a default value when the arguments are omitted.

**Comment by HertzDevil** [comment]
The default arguments are arguments provided at the method definition site rather than at the invocation site. I don't see why these expressions aren't arguments.

**Comment by straight-shoota** [comment]
They could be seen as arguments, yes. But the values are assigned at the parameter definition. To me it seems less confusing to just talk about values than to pick on the difference between parameters and arguments in this case.

**Comment by Fryguy** [comment]
"default argument" is a weird turn of phrase.  It's not incorrect, but "default value" feels so much more common.  "default argument values for the last parameters" is probably the most technically accurate, but that feels a bit verbose.  I think "value" and "argument" might be interchangable?

**Comment by HertzDevil** [comment]
So I tried to gather the reference manuals of a bunch of programming languages to see which terms they use:

* [C++](https://en.cppreference.com/w/cpp/language/default_arguments): default argument
* [C#](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/language-specification/classes#method-parameters): default argument
* [D](https://dlang.org/spec/function.html#function-default-args): default argument, default value
* [Dart](https://dart.dev/guides/language/specifications/DartLangSpec-v2.2.pdf#subsubsection.9.2.2): default value
* [Delphi](http://docwiki.embarcadero.com/RADStudio/Sydney/en/Parameters_(Delphi)#Default_Parameters): default parameter value
* [Elixir](https://elixir-lang.org/getting-started/modules-and-functions.html#default-arguments): default argument, default value
* [Javascript](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/Default_parameters): default argument, default parameter value
* [Kotlin](https://kotlinlang.org/docs/reference/functions.html#default-arguments): default argument, default parameter value
* [PHP](https://www.php.net/manual/en/functions.arguments.php#functions.arguments.default): default argument, default value
* [Python](https://docs.python.org/3/reference/compound_stmts.html#function-definitions): default parameter value
* [Raku](https://design.raku.org/S06.html#Optional_parameters): default value
* [Ruby](https://ruby-doc.org/core-3.0.0/doc/syntax/methods_rdoc.html#label-Default+Values): default value
* [Scala](https://scala-lang.org/files/archive/spec/2.13/04-basic-declarations-and-definitions.html#default-arguments): default argument
* [Swift](https://docs.swift.org/swift-book/ReferenceManual/Declarations.html#ID472): default value (default-argument-clause in the grammar)
* Java, Lua, Objective-C, Rust: unsupported

"Default argument value" is redundant; the value (or the expression) _is_ the argument.

**Comment by RespiteSage** [comment]
Though it's a bit wordy, I think "default parameter value" strikes a good balance between accuracy and understandability.

**Comment by HertzDevil** [comment]
Judging from those languages it seems Ruby's documentation is ironically the loosest one with regard to "arguments" and "parameters", though the [Japanese reference](https://docs.ruby-lang.org/ja/3.0.0/doc/spec=2fdef.html#method) is better:

> 仮引数にデフォルト式が与えられた場合、メソッド呼び出しで実引数を省略したときのデフォルト値になります。

*"If a default expression is provided for a formal argument, it becomes that argument's default value when the actual argument is omitted in a method call."*

**Comment by oprypin** [comment]
Slight preference for "Default value" for me.

I can understand "default argument" as "argument that's being passed if nothing else is". But an argument is something that a user of the function actively passes, by writing it into parentheses, so to say. But here it's kind of always been there, you don't need to pass it.

It's certainly not wrong, just sounds weird indeed.

In Python, though... oh, this would definitely be wrong. Python has actual default *values* that just sit there attached to the function as a (mutable!) singleton.

**Comment by potomak** [comment]
I noticed an incorrect usage use of arg/param terms in the parser’s source code too.

I would suggest to update all variable names used to store param names and values from ‘arg_*’ to ‘param_*’.

Examples:
* block params 
* block grouped params
* function definition params
* macro definition params

**Comment by straight-shoota** [comment]
For going all the way, we should probably also rename the `Crystal::Arg` type to `Crystal::Param`. I'm not sure if that would have any implications that we want to avoid. Leaving `Arg` in as an alias should probably be good for tools that hook directly into the compiler API (for example ameba).

**Comment by potomak** [comment]
I was thinking the same and I was also wandering if Crystal has aliases for keeping the backward compatibility. I can publish a PR for renaming the class.

There are also a couple of struct fields that should be renamed.
